pkg/network/prober/prober.go

```diff
@@ -26,40 +26,70 @@ import (
 	"github.com/pkg/errors"
 	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/apimachinery/pkg/util/wait"
-
-	"github.com/knative/serving/pkg/network"
 )
 
 // TransportFactory is a function which returns an HTTP transport.
 type TransportFactory func() http.RoundTripper
 
-// ProbeOption is a way for caller to modify the HTTP request before it goes out.
-type ProbeOption func(r *http.Request) *http.Request
+// Option provides a way for the caller to mutate the HTTP request before it goes out,
+// and to validate the HTTP response after it comes back.
+type Option struct {
```

Comment:
I think those are two different types, but I am not hard set on this.

Reply:
They are, but then there is no way to have the a clean variadic function without using interface{}.
The 3 options are:
- current option: statically typed, clean variadic function
- 2 types, interface{} variadic function: not statically typed
- 2 types, statically typed: Do becomes something like Do(..., []RequestOption, []ValidateOption)

Reply:
We do usually []interface{} and then to type switch.
It happens in at least two places in the code.

Reply:
interface{} is disgusting, but I'll change it to two types and interface.

Reply:
it totally is ;-)